### PR TITLE
[SPARK-47318][CORE][3.4] Adds HKDF round to AuthEngine key derivation to follow standard KEX practices

### PR DESCRIPTION
**Pull request by sweisdb**
### What changes were proposed in this pull request?

Backport of SPARK-47318 to v3.4.0

This change adds an additional pass through a key derivation function (KDF) to the key exchange protocol in `AuthEngine`. Currently, it uses the shared secret from a bespoke key negotiation protocol directly. This is an encoded X coordinate on the X25519 curve. It is atypical and not recommended to use that coordinate directly as a key, but rather to pass it to an KDF.

Note, Spark now supports TLS for RPC calls. It is preferable to use that rather than the bespoke AES RPC encryption implemented by `AuthEngine` and `TransportCipher`.

### Why are the changes needed?

This follows best practices of key negotiation protocols. The encoded X coordinate is not guaranteed to be uniformly distributed over the 32-byte key space. Rather, we pass it through a HKDF function to map it uniformly to a 16-byte key space.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Exiting tests under:
`build/sbt "network-common/test:testOnly"`

Specifically:
`build/sbt "network-common/test:testOnly org.apache.spark.network.crypto.AuthEngineSuite"`
`build/sbt "network-common/test:testOnly org.apache.spark.network.crypto.AuthIntegrationSuite"`

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: backport.patch [hunk 1]
```diff
@@ -0,0 +1,219 @@
+diff --git a/common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java b/common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java
+index 078d9ceb31..14f0c23fd0 100644
+--- a/common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java
++++ b/common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java
+@@ -41,16 +41,19 @@ import org.apache.spark.network.util.TransportConf;
+  * Exchange, using a pre-shared key to derive an AES-GCM key encrypting key.
+  */
```
Comment (mridulm) [inline backport.patch:7]: Remove this file ?

File: docs/security.md.orig [hunk 2]
```diff
@@ -0,0 +1,960 @@
...
+title: Security
+license: |
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+ 
+     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (mridulm) [inline docs/security.md.orig:13]: Remove this file as well ?

**Review by mridulm — COMMENTED** [review]
Btw, if #46014 can cleanly apply to 3.4 branch as well, we dont need this PR.
Also, same query as https://github.com/apache/spark/pull/46014/files#r1561562093 applies here as well

**Comment by sweisdb** [comment]
> Btw, if #46014 can cleanly apply to 3.4 branch as well, we dont need this PR. Also, same query as https://github.com/apache/spark/pull/46014/files#r1561562093 applies here as well

Yes, I think 46014 applies cleanly here. Happy to close this PR if we can merge that to multiple branches.

**Comment by dongjoon-hyun** [comment]
To @mridulm , I prefer to pass CI. :)

**Review by dongjoon-hyun — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Merged to branch-3.4 for Apache Spark 3.4.3.

**Comment by mridulm** [comment]
Thanks @dongjoon-hyun !

**Comment by dongjoon-hyun** [comment]
Thank you, @mridulm .

**Pull request closed without merging**